WIP: Add descriptions to register_block_pattern.

carolinan opened the pull request:
## Description
I have attempted to add descriptions to block patterns.
Fixes #23042

I am also not sure if I got the ID right.

`const descriptionId = `block-editor-inserter__patterns-item-description-${ instanceId }`;`

Output: 
`<div class="components-visually-hidden" id="block-editor-inserter__patterns-item-description-4">A cover block with a gradient background and a paragraph with an extra large font size.</div>`

## How has this been tested?
I added descriptions to four of the existing patterns, as well as a pattern inside a theme.
I tested it by checking the markup on the inserter, and tested the aria described by and visually hidden element with the NVDA screen reader.

## Types of changes
New feature.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

carolinan committed:
Merge pull request #1 from WordPress/master

sync

carolinan committed:
Merge pull request #3 from WordPress/master

sync

carolinan committed:
Merge pull request #5 from WordPress/master

sync

carolinan committed:
Add description to block patterns

Adds visually hidden descriptions to block patterns in the inserter. For #23042.

youknowriad commented on packages/block-editor/src/components/inserter/block-patterns.js:
```diff
@@ -37,11 +41,15 @@ function BlockPattern( { pattern, onClick } ) {
 			<div className="block-editor-inserter__patterns-item-title">
 				{ pattern.title }
 			</div>
+			<VisuallyHidden id={ descriptionId }>
```
you can add 

```
{ !! pattern.description && (
// Your visibily hidden element
} )
```

to render this conditionally depending on the presence of the description.

youknowriad reviewed (commented):
This is looking good to me. I'd appreciate @enriquesanchez's review

carolinan committed:
Merge pull request #6 from WordPress/master

sync with master

carolinan committed:
Merge pull request #7 from carolinan/master

sync branch with master

carolinan commented:
How do you all feel about renaming "title" to label, matching register_block_style? Does it need to be a separate issue?

youknowriad commented:
We do have "title" for blocks too, so it seems the inconsistency already exists.

carolinan committed:
Add doing it wrong messages

Adds doing_it_wrong messages for the required properties.
Adds descriptions to the block patterns.
Updates documentation with the description and viewportWidth.

youknowriad commented on docs/designers-developers/developers/block-api/block-patterns.md:
```diff
@@ -15,16 +15,19 @@ The `register_block_pattern` function receives the name of the pattern as the fi
 The properties of the pattern include:
  - `title` (required): A human-readable title for the pattern.
  - `content` (required): Raw HTML content for the pattern.
+ - `description` (required): A description of the pattern. Used to describe the pattern in the inserter.
```
I'm personally hesitant about making this required. I think it shouldn't be as a lot of existing patterns don't have it and we should fallback to the title. Sometimes the title is descriptive enough.

enriquesanchez replied:
Is there a way to not have it be required but be really really encouraged?

youknowriad replied:
mmm I don't know, we could console log a message if we don't find it maybe but we never did that before and it can be very verbose. I can't think of a way though.

carolinan replied:
Is compatibility for existing patterns the only concern? _Because the function was just renamed,_ developers had to make changes then, why is adding a description a bigger concern?

youknowriad replied:
I've thought about this more and I still think it shouldn't be mandatory. Some patterns are descriptive enough and the API should be as straightforward to use as possible.

Enforcing on a potential Patterns directory/repository is a separate question though and I'd see this being different there.

carolinan replied:
But I thought that was the point, the patterns are not descriptive if you can't see them.
And currently, neither the title or the content is required, even though the documentation says they are.
Without a title, the only thing announced by screen readers is "button".

Adding a description is more straight forward than adding categories (especially if you throw in registering the categories).

enriquesanchez commented:
This is a great PR, @carolinan! Tested with VoiceOver + Safari and works as expected. The descriptions you added are very detailed and it's easy to picture the pattern.

carolinan committed:
Merge branch 'master' into update/block-patterns

youknowriad commented:
How can we move this PR forward to include in 5.5

enriquesanchez commented:
It'll be great to have this included with 5.5 and updated to reflect the changes in #23608.

Anything left to be done before we can merge?

youknowriad commented:
I'd personally like for the mandatory aspect to be removed to include this as stated here https://github.com/WordPress/gutenberg/pull/23070#discussion_r444944151 especially since we're late in the 5.5 process and people are likely to have created a lot of patterns by now.

carolinan commented:
So to summarize the discussion during the accessibility teams meeting on friday:
Title will be required.
Description will be strongly recommended.

https://github.com/WordPress/gutenberg/pull/23608 needs to be merged first so that the list of patterns is up to date.

carolinan committed:
Merge pull request #8 from WordPress/master

sync with master

carolinan committed:
trying to solve merge conflicts

trying to solve merge conflicts with the inserter updates

carolinan committed:
Merge pull request #9 from carolinan/master

sync with master

carolinan committed:
No longer require descriptions

Removes the check for the description property in the pattern registry.

Update the documentation to show that descriptions are encouraged, not required.

carolinan committed:
Add the description to the block patterns list

Adds the description to the block patterns list.
Outputs the visually hidden description after the title, if the description exists.

carolinan committed:
Update block-patterns.md

Updated the wording in the description for the viewportWidth.

youknowriad commented:
@carolinan That sounds great 👍 

#23608 is merged, do you have time to refresh the PR to include in today's RC?

carolinan committed:
remove the deleted block patterns

carolinan committed:
Text update to solve merge conflict

carolinan committed:
remove description

carolinan committed:
update pattern content

carolinan committed:
Merge pull request #11 from WordPress/master

sync branch with master

youknowriad commented on lib/class-wp-block-patterns-registry.php:
```diff
@@ -42,6 +42,18 @@ public function register( $pattern_name, $pattern_properties ) {
+			$message = __( 'Pattern content must be a string.', 'gutenberg' );
+			_doing_it_wrong( __METHOD__, $message, '8.5.0' );
+			return false;
+		}
```
The change here needs dedicated Core patch.

carolinan replied:
I am not sure where to upload it :) https://core.trac.wordpress.org/attachment/ticket/50550/50550-doing-it-wrong.patch

youknowriad reviewed (approved):
Looks great. Thanks @carolinan

youknowriad commented:
I merged too quickly without rewording the commit message. I guess it's fine for this time :).
